README.md
[![Build Status](https://travis-ci.org/gismo141/homebridge-server.svg?branch=master)](https://travis-ci.org/gismo141/homebridge-server)

[![NPM](https://nodei.co/npm/homebridge-server.png?downloads=true&downloadRank=true)](https://nodei.co/npm/homebridge-server/)
[![NPM](https://nodei.co/npm-dl/homebridge-server.png?&months=6&height=3)](https://nodei.co/npm/homebridge-server/)

# Homebridge Server - a plugin to configure your Homebridge-devices in your Browser

The purpose of this plugin is to change your homebridge configuration via the webbrowser.
The plugin handles the JSON-part and you - as a user - can add your devices to the bridge.

![Overview](resources/overview.png)

See my blog post [Configure your Homebridge with Homebridge-Server](https://gismo141.github.io/configure-your-homebridge-2/) on how I implemented this plugin.

## What can you do with this plugin?

1. Change the broadcasted name of your Homebridge
2. Change the MAC-address used to identify your Homebridge
3. Change the PIN to verify your Homebridge
4. Add or Remove platforms
5. Add or Remove accessories
6. Change the names of your services
7. Backup your configuration
8. Show the Log-file of Homebridge
9. Reboot your system

## How to setup?

Fast way:

```Bash
[sudo] npm install homebridge-server@latest -g
```

Add this snippet to your `config.json`:

```JSON
{
    "platform": "Server",
    "port"    : 8765,
    "name"    : "Homebridge Server",
    "log"     : "<PATH_TO_YOUR_LOG_FILE>",
    "restart" : "<COMMAND_TO_EXECUTE>"
}
```

The parameter `"log"` defines where the plugin can find your log.
If you're using `systemd` to execute `homebridge` then `<PATH_TO_YOUR_LOG_FILE>` should be `"systemd"`.
If you're executing `homebridge` from a script, then you should pipe its output to a file.
In this case `<PATH_TO_YOUR_LOG_FILE>` should be the path to the piped output.

and launch or restart your homebridge, e.g.:

```Bash
homebridge
```

See the [Wiki](https://github.com/gismo141/homebridge-server/wiki) on how to configure, install and use the plugin for more complex scenarios.

## Backup of your Configuration

If you want to backup your configuration, hit the Backup-button in the menubar of the webpage.
The plugin saves the configuration at the same location of your original configuration but appends an `.bak`-extension.
Whenever you crash your config you should have a valid backup!

![Backup](resources/backup.png)

## Log-File

You can directly see what's happening on your Homebridge - hit the button in the menubar and the webserver shows you the log in your browser!

![Log File](resources/log_file.png)

## Manage your Homebridge-plugins

From this version you are now able to browser, remove, and install all Homebridge-plugins that are available on [https://npmjs.com](https://npmjs.com)!
Thanks to the great API of [https://npms.io](https://npms.io) you can filter the plugins by name and description.

![Plugins](resources/plugins.png)

**At the moment you can receive a 250 results. I'll fix it as soon as I can! ;)**

## Rebooting your System

To make the changes in your configuration permanent, you need to restart the Homebridge-service.
If you want, you can restart your Homebridge-running system directly via the webpages `Reboot`-button.
Hit the button in the menubar and wait, until your system has rebooted and your services are back online.

**Disclaimer**

The reboot is immediate!
If your Homebridge is not running as a service but as a standalone program you launch manually, the Homebridge won't restart automatically.
You must setup Homebridge as a service on boot - This is necessary for the plugin to work!

Please make sure to restart Homebridge manually if you haven't set it up as a service.

If you want to start homebridge as a service, see [Running Homebridge on Bootup](https://github.com/nfarina/homebridge/wiki/Running-Homebridge-on-a-Raspberry-Pi#running-homebridge-on-bootup).

## How could you contribute?

1. [Fork this project][fork] to your account.
2. [Create a new branch][branch] for the improvements, you intend to make.
3. **Make the changements in your fork.**
4. [Send a pull-request][pr] from your fork’s branch to my `master` branch.

You can always use the web-interface to make the changes you want. It helps you automizing the workflow from above.

[fork]: http://help.github.com/forking/
[branch]: https://help.github.com/articles/creating-and-deleting-branches-within-your-repository
[pr]: http://help.github.com/pull-requests/


Changelog.md
## 2.0.0
- Refactoring
- Config option for path of node_modules; if not set, '/usr/local/lib/node_modules/' is assumed.
- Extended JSON-API
- Main view
    - Show info about homebridge instance:
        - version (installed and last release)
        - uptime
        - memory usage
        - host OS
    - Indicate if platform is active (or just configured)
- Plugins view
    - Plugin update, remove and install:
        - Check access rights before
        - Show progress during plugin update, remove and install
    - Installed plugins
        - Show version info
        - Show if plugin is currently used by any platform or accessory
        - Update plugins
    - Search plugins
        - Live search
        - Indicate compatibility
        - Mark plugins that are already installed


## TODO:

### Main view

- [x] Provide link to plugin homepage in platforms and accessories list
- [ ] Use icons for platform/accessory status (running, warning, stopped)
- [ ] Add 'description' to bridgeInfo
- [ ] Plugins can provide an example config in their package; use it for a ConfigWizard

### Plugins

- [ ] Installed plugins
  - [x] Show release date info for installed plugins with update
  - [ ] Show link if plugin is not configured
- [ ] Operation
  - [ ] Highlight if error occured.

### API

- [ ] Convert backend error codes to helpful text (e.g. 'EACCES')
- [ ] restartHomebridge: rework GUI; current approach doesn't work since the server is terminated...

### PluginManager

- [x] getInstalledPlugins: derive modulePath from config.modulePath


api/PluginManager.js
/* eslint-env node */

'use strict';

module.exports = {
  PluginManager: PluginManager
}


// Internals

// _plugins holds an array with information about each installed plugin.
// It has the following properties:
// - name : The name of the plugin; e.g. 'xxx'
// - description : description
// - author : The username of the plugins publisher
// - homepage : URL of the plugins homepage
// - version : The version; derived from the plugins package.json
// - latestVersion : The latest version published on npmjs.io
// - isLatestVersion : Flag describing if the installed version is the latest available; Values: "n/a", "1" or "0".
// - homebridgeMinVersion : The minimum version of homebridge the plugin needs
// - platformUsage : Number of platforms using this plugin
// - accessoryUsage : Number of accessories using this plugin
var _plugins = {};

var hbsPath = "";
var hbLog = function() {};       // eslint-disable-line

function PluginManager(libPath, log) {
    hbsPath = libPath;
    hbLog = log;
    getInstalledPlugins();
}

PluginManager.prototype.plugins = function() {
    return _plugins;
}


PluginManager.prototype.search = function(query, callback) {
    var path = require(path);
    var utilsLib = require(path.resolve(hbsPath, 'api', 'utils.js'));
    var utils = new utilsLib.Utils();

    var options = {
        host: 'api.npms.io',
        port: 443,
        path: '/v2/search?q=' + (!query || 0 === query.length ? '' : query + '+') + 'keywords:homebridge-plugin+not:deprecated+not:insecure&size=250',
        method: 'GET',
        headers: {
            'Content-Type': 'application/json'
        }
    };
    utils.getJSON(options, function(statusCode, result) {
        var results = result.results;

        // Check for each result if already installed
        // TODO: this is inefficient for sites with a lot of plugins...
        for (var result_id in results) {
            results[result_id].hb_IsInstalled = 0;
            for (var pluginID in _plugins) {
                var plugin = _plugins[pluginID];
                if (plugin.name.toLowerCase().replace(/ /g, '-') === results[result_id].package.name) {
                    results[result_id].hb_IsInstalled = 1;
                    // results[pf_ID].hb_installedVersion = pf.version;
                }
            }
        }
        callback(results);
    });
}

function npmOperation(options, callback) {
    var spawn = require('child_process').spawn;
    var npmProcess = spawn('npm', options);

    npmProcess.stdout.on('data', function (data) {
        callback(false, data, false);
    });

    npmProcess.stderr.on('data', function (data) {
        callback(false, data, false);
    });

    npmProcess.on('close', function (code) {
        callback(true, code, true);
    });
}

PluginManager.prototype.installPlugin = function(pluginName, callback) {
    checkAccessRights(pluginName, function(result, msg) {
        if (result === false) {
            callback(false, msg, true);
            return;
        }
        // npmOperation(['install', '-g', pluginName, '--dry-run'], callback);
        npmOperation(['install', '-g', pluginName], callback);
    });
}


PluginManager.prototype.updatePlugin = function(pluginName, callback) {
    checkAccessRights(pluginName, function(result, msg) {
        if (result === false) {
            callback(false, msg, true);
            return;
        }
        // We use 'npm install' instead of 'npm update' since 'update' ignores the version tag.
        // npmOperation(['install', '-g', pluginName, '--dry-run'], callback);
        npmOperation(['install', '-g', pluginName], callback);
    });
}


PluginManager.prototype.removePlugin = function(pluginName, callback) {
    checkAccessRights(pluginName, function(result, msg) {
        if (result === false) {
            callback(false, msg, true);
            return;
        }
        // npmOperation(['uninstall', '-g', pluginName, '--dry-run'], callback);
        npmOperation(['uninstall', '-g', pluginName], callback);
    });
}

/**
 * Scans the node-modules directory for homebridge plugins.
 * Subsequently calls enrichUsageInfo() and enrichMetadata() to gather further information.
 * _plugins gets updated by these functions.
 */
function getInstalledPlugins() {
    var fs = require('fs');
    var path = require('path');
    var plugins = [];

    var globalModulePath = require('global-modules') + "/";
    var possiblePaths = [globalModulePath, hbsPath + "../"];
    possiblePaths.forEach(function(modulePath) {
        if (! fs.existsSync(path.normalize(modulePath))) {
            return;
        }
        var modules = fs.readdirSync(path.normalize(modulePath));
        for (var moduleID in modules) {
            var moduleName = modules[moduleID];
            if (moduleName.startsWith('homebridge-')) {
                var packagePath = modulePath + moduleName + "/package.json";
                var packageJSON = "";
                try {
                  packageJSON = require(packagePath);
                  var plugin = {
                      "name": moduleName,
                      "version": packageJSON.version,
                      "latestVersion": "n/a",
                      "isLatestVersion": "n/a",
                      "platformUsage": 0,
                      "accessoryUsage": 0
                  }
                  plugins.push(plugin);
                } catch (e) {
                  hbLog("-> PluginManager:getInstalledPlugins(): " + moduleName + " will be ignored because it is not a valid homebridge plugin.");
                }
            }
        }
    });
    _plugins = plugins;
    enrichUsageInfo();
    enrichMetadata();
}


/**
 * Set information if a plugin is used by this homebridge site.
 * Updates _plugins with this information.
 */
function enrichUsageInfo() {
    var utilsLib = require(hbsPath + 'api/utils.js');
    var utils = new utilsLib.Utils();

    var options = {
        // host: 'api.npms.io',
        port: 8765,
        path: '/api/installedPlatforms',
        method: 'GET',
        headers: {
            'Content-Type': 'application/json'
        }
    };
    utils.getJSON(options, function(statusCode, platforms) {
        for (var pluginID in _plugins) {
            for (var pf_id in platforms) {
                if (platforms[pf_id]['hbServer_pluginName'] === _plugins[pluginID]["name"]) {
                    _plugins[pluginID]['platformUsage']++;
                }
            }
        }
    });
}


/**
 * Retrieve further information from npms.io (by calling fetchPluginMetaData()) and store it in the plugin properties.
 * Updates _plugins.
 */
function enrichMetadata() {
    var semver = require('semver');

    for (var pluginID in _plugins) {
        fetchPluginMetaData(_plugins[pluginID].name, pluginID, function(metadata, cpluginID) {
            _plugins[cpluginID]["description"] = "n/a";
            _plugins[cpluginID]["author"] = "n/a";
            _plugins[cpluginID]["homepage"] = "n/a";
            _plugins[cpluginID]["homebridgeMinVersion"] = "n/a";
            _plugins[cpluginID]["latestVersion"] = "n/a";
            _plugins[cpluginID]["isLatestVersion"] = "n/a";
            if  (metadata["collected"].hasOwnProperty("metadata")) {
                _plugins[cpluginID]["description"] = metadata["collected"]["metadata"].hasOwnProperty("description") ? metadata["collected"]["metadata"]["description"] : "n/a";

                _plugins[cpluginID]["author"] = "n/a";
                if (metadata["collected"]["metadata"].hasOwnProperty("publisher")) {
                    _plugins[cpluginID]["author"] = metadata["collected"]["metadata"]["publisher"].hasOwnProperty("username") ? metadata["collected"]["metadata"]["publisher"]["username"] : "n/a";
                }

                _plugins[cpluginID]["homepage"] = "n/a";
                if (metadata["collected"]["metadata"].hasOwnProperty("links")) {
                    _plugins[cpluginID]["homepage"] = metadata["collected"]["metadata"]["links"].hasOwnProperty("homepage") ? metadata["collected"]["metadata"]["links"]["homepage"] : "n/a";
                }

                _plugins[cpluginID]["homebridgeMinVersion"] = "n/a";
                if (metadata["collected"]["metadata"].hasOwnProperty("dependencies")) {
                    _plugins[cpluginID]["homebridgeMinVersion"] = metadata["collected"]["metadata"]["dependencies"].hasOwnProperty("homebridge") ? metadata["collected"]["metadata"]["dependencies"]["homebridge"] : "n/a";
                }

                var latestVersion = metadata["collected"]["metadata"].hasOwnProperty("version") ? metadata["collected"]["metadata"]["version"] : "n/a";
                _plugins[cpluginID]["latestVersion"] = latestVersion;
                if (semver.lt(_plugins[cpluginID]["version"], latestVersion)) {
                    _plugins[cpluginID]["isLatestVersion"] = "0";
                } else {
                    _plugins[cpluginID]["isLatestVersion"] = "1";
                }
            }
        });
    }
}


/**
 * Queries npms.io for plugin metadata. Used by enrichMetadata().
 * @param  {string}   pluginName   Name of the plugin to query
 * @param  {Function} callback Called with the query result.
 */
function fetchPluginMetaData(pluginName, pluginID, callback) {
    var utilsLib = require(hbsPath + 'api/utils.js');
    var utils = new utilsLib.Utils();

    var searchname = (pluginName.toLowerCase().replace(/ /g, '-'));

    var options = {
        host: 'api.npms.io',
        port: 443,
        path: '/v2/package/' + searchname,
        method: 'GET',
        headers: {
            'Content-Type': 'application/json'
        }
    };
    utils.getJSON(options, function(statusCode, result) {
        if (statusCode === 200) {
            callback(result, pluginID);
        }
    });
}


function checkAccessRights(pluginName, callback) {
    var fs = require('fs');
    var globalModulePath = require('global-modules') + "/";
    var modulePath = globalModulePath;

    var pluginNameClean = pluginName;

    // Remove version tag if provided
    var tagIndex = pluginName.indexOf("@");
    if (tagIndex !== -1) {
        pluginNameClean = pluginName.substr(0, tagIndex);
    }

    // first, check if global node_modules directory is read/writeable
    fs.access(modulePath, fs.R_OK | fs.W_OK, function(err) {
        if (err) {
            callback(false, "EACCES: No rights to write in " + modulePath);
            return;
        }

        // second, check the plugin directory...
        modulePath = globalModulePath + pluginNameClean;
        // ... if it already exists...
        if (fs.existsSync(modulePath) === true) {
            // ... it must be writeable
            fs.access(modulePath, fs.R_OK | fs.W_OK, function(err) {
                if (err) {
                    callback(false, "EACCES: No rights to write in " + modulePath);
                    return;
                }
                callback(true, "ok");
                return;
            });
        } else {
            callback(true, "ok");
        }
    });
}
